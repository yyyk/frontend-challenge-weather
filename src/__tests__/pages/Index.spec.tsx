import { render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import MOCK_DATA from 'src/__tests__/mock/openWeatherMapData.json';
import SelectedWeatherProvider from 'src/providers/SelectedWeather';
import IndexPage from 'src/pages/Index';

jest.mock('src/components/Header', () => () => <div data-testid="header" />);
jest.mock('src/components/Card', () => () => <div data-testid="card" />);

describe('<IndexPage> (loading)', () => {
  beforeEach(() => {
    jest.spyOn(global, 'fetch').mockResolvedValue({
      json: jest.fn().mockResolvedValue(null),
    } as any);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('matches snapshot', async () => {
    let asFragment;
    await act(async () => {
      const result = await render(
        <SelectedWeatherProvider>
          <IndexPage />
        </SelectedWeatherProvider>,
      );
      asFragment = result.asFragment;
    });
    expect((asFragment as any)()).toMatchSnapshot();
  });

  it('displays loading message', async () => {
    const message = 'Loading...';
    await act(async () => {
      await render(
        <SelectedWeatherProvider>
          <IndexPage />
        </SelectedWeatherProvider>,
      );
    });
    expect(screen.getByText(message)).toBeTruthy();
  });
});

describe('<IndexPage> (loaded)', () => {
  beforeEach(() => {
    jest.spyOn(global, 'fetch').mockResolvedValue({
      json: jest.fn().mockResolvedValue(MOCK_DATA),
    } as any);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('matches snapshot', async () => {
    let asFragment;
    await act(async () => {
      const result = await render(
        <SelectedWeatherProvider>
          <IndexPage />
        </SelectedWeatherProvider>,
      );
      asFragment = result.asFragment;
    });
    expect((asFragment as any)()).toMatchSnapshot();
  });

  it('renders Header component', async () => {
    await act(async () => {
      await render(
        <SelectedWeatherProvider>
          <IndexPage />
        </SelectedWeatherProvider>,
      );
    });
    expect(screen.getByTestId(/header/)).toBeInTheDocument();
  });

  it('renders Card component', async () => {
    await act(async () => {
      await render(
        <SelectedWeatherProvider>
          <IndexPage />
        </SelectedWeatherProvider>,
      );
    });
    expect(screen.getAllByTestId(/card/).length).toBeGreaterThan(0);
  });
});
