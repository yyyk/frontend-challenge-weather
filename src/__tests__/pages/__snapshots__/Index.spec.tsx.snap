// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexPage> (loaded) matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="css-vsj5tt"
  >
    <div
      class="css-12dakew"
    >
      <div
        data-testid="header"
      />
      <ul
        class="css-1yotrj6"
      >
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
        <li
          class="css-1u62lmq"
        >
          <div
            data-testid="card"
          />
        </li>
      </ul>
    </div>
  </main>
</DocumentFragment>
`;

exports[`<IndexPage> (loading) matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="css-vsj5tt"
  >
    <div
      class="css-1hfhcgr"
    >
      Loading...
    </div>
  </main>
</DocumentFragment>
`;
